src/gtc/passes/oir_optimizations/utils.py

```diff
@@ -201,35 +202,59 @@ class Context:
         fields: Dict[str, Extent] = field(default_factory=dict)
         blocks: Dict[int, Extent] = field(default_factory=dict)
 
+    def __init__(self, add_k: bool = False):
+        self.add_k = add_k
+        self.zero_extent = Extent.zeros(ndims=2)
+
     def visit_Stencil(self, node: oir.Stencil) -> "Context":
         ctx = self.Context()
         for vloop in reversed(node.vertical_loops):
             self.visit(vloop, ctx=ctx)
 
+        if self.add_k:
+            ctx.fields = {name: Extent(*extent, (0, 0)) for name, extent in ctx.fields.items()}
+
         return ctx
 
     def visit_VerticalLoopSection(self, node: oir.VerticalLoopSection, **kwargs: Any) -> None:
         for hexec in reversed(node.horizontal_executions):
             self.visit(hexec, **kwargs)
 
     def visit_HorizontalExecution(self, node: oir.HorizontalExecution, *, ctx: Context) -> None:
-        results = AccessCollector.apply(node).cartesian_accesses()
+        results = AccessCollector.apply(node)
         horizontal_extent = functools.reduce(
-            lambda ext, name: ext | ctx.fields.get(name, Extent.zeros(ndims=2)),
+            lambda ext, name: ext | ctx.fields.get(name, self.zero_extent),
             results.write_fields(),
-            Extent.zeros(ndims=2),
+            self.zero_extent,
         )
         ctx.blocks[id(node)] = horizontal_extent
 
         for name, accesses in results.read_offsets().items():
             extent = functools.reduce(
-                lambda ext, off: ext | Extent.from_offset(off[:2]), accesses, Extent.zeros(ndims=2)
-            )
-            ctx.fields[name] = ctx.fields.get(name, Extent.zeros(ndims=2)).union(
-                horizontal_extent + extent
+                lambda ext, off: ext | Extent.from_offset(off[:2]),
+                accesses,
+                Extent.from_offset(accesses.pop()[:2]),
             )
+            total_extent = horizontal_extent + extent
+            ctx.fields.setdefault(name, total_extent)
+            ctx.fields[name] |= total_extent
+
+        for name in results.write_fields():
+            ctx.fields.setdefault(name, horizontal_extent)
 
 
-def compute_horizontal_block_extents(node: oir.Stencil) -> Dict[int, Extent]:
-    ctx = StencilExtentComputer().visit(node)
+def compute_horizontal_block_extents(node: oir.Stencil, **kwargs: Any) -> Dict[int, Extent]:
+    ctx = StencilExtentComputer(**kwargs).visit(node)
```

Comment:
It doesn't have to be in this PR, but I can't find in the stack of horizontal region PR's where this gets extended to the case of regions. Did I miss that?

Reply:
Ah ha! That means the change is well hidden in layers of abstraction. The specific change is separating the concepts of offsets and extents. There is a method that converts between them given the extent for the horizontal execution (`horizontal_extent`). See [here](https://github.com/GridTools/gt4py/pull/678/files#diff-1ad76d78dea66f59e27bce7379282bc29ce0633d4fa3274c1b9773882d4bcfd7R252) for the call.

Reply:
:+1: thanks for the explanation